docx/document.py

```diff
@@ -237,6 +250,39 @@ def _body(self):
             self.__body = _Body(self._element.body, self)
         return self.__body
 
+    def _calculate_next_footnote_reference_id(self, p):
+        """
+        Return the appropriate footnote reference id number for
+        a new footnote added at the end of paragraph `p`.
+        """
+        # When adding a footnote it can be inserted 
+        # in front of some other footnotes, so
+        # we need to sort footnotes by `footnote_reference_id`
+        # in |Footnotes| and in |Paragraph|
+        new_fr_id = 1
+        # If paragraph already contains footnotes
+        # and it's the last paragraph with footnotes, then
+        # append the new footnote and the end with the next reference id.
+        if p.footnote_reference_ids is not None:
+            new_fr_id = p.footnote_reference_ids[-1] + 1
+        # If the document has footnotes after this paragraph,
+        # the increment all footnotes pass this paragraph, 
+        # and insert a new footnote at the proper position.
+        # break the loop when we get to the footnote before the one we are inserting.
+        has_passed_self = False
```

Comment:
Now self is document, so this is a bit misleading in this scenario.
Maybe something like `has_passed_containing_para`?
It's a bit verbose, but it's the first thing that came to my mind.

Reply:
Agree